include/xsimd/types/xsimd_traits.hpp

```diff
@@ -53,6 +54,21 @@
 namespace xsimd
 {
 
+    namespace detail
+    {
+        // Forward define for later use in this file
+        template <class T>
+        struct is_complex;
```

Comment:
That should not be required, `is_complex` is defined in `xsimd_complex_base.hpp`  which is indirectly included in `xsimd_types_include.hpp` which is included above. So the fix should be to include any missing file defining complex batch in `xsimd_types_include.hpp`.

Reply:
Wouldn't this eventually break for builds without SIMD then?
Shouldn't is_complex be in a different always-included utils file?

I'll take another look in a bit.

Reply:
Builds without SIMD should not include any file from `xsimd.hpp` (either because the architecture is too old or because we don't support the instruction set).

Now your remark is relevant, since we're using this metafunction out of batch related stuff, we may move it to the `xsimd_utils.hpp` file.